tuning/data/data_processors.py

```diff
@@ -153,19 +174,41 @@ def _process_dataset_configs(
 
                     raw_datasets = raw_datasets.map(handler, **kwargs)
 
-            if final_datasets is None:
-                final_datasets = raw_datasets
-            else:
-                for k in raw_datasets.keys():
-                    if k in final_datasets:
-                        final_datasets[k] = datasets.concatenate_datasets(
-                            [final_datasets[k], raw_datasets[k]]
-                        )
-                    else:
-                        final_datasets[k] = raw_datasets[k]
-
-        if "train" in final_datasets:
-            train_dataset = final_datasets["train"]
+            # Append the processed datasets to the final dict
+            all_datasetdicts.append(raw_datasets)
+
+        # This is a dict of { split: list[datasets] }
+        final_datasets = {}
+        for d in all_datasetdicts:
+            for k, v in d.items():
+                if k not in final_datasets:
+                    final_datasets[k] = [v]
+                else:
+                    final_datasets[k].append(v)
+
+        if sample_datasets:
+            strategy = self.processor_config.sampling_stopping_strategy
+            seed = self.processor_config.sampling_seed
+            logging.info(
+                "Interleaving datasets: strategy[%s] seed[%d] probabilities[%s]",
+                strategy,
+                seed,
+                str(sampling_probabilities),
+            )
+            for k, v in final_datasets.items():
+                interleaved = datasets.interleave_datasets(
+                    datasets=v,
+                    probabilities=sampling_probabilities,
+                    stopping_strategy=strategy,
+                    seed=seed,
+                )
+                final_datasets[k] = interleaved
+        else:
+            for k, v in final_datasets.items():
+                combined = datasets.concatenate_datasets(v)
+                final_datasets[k] = combined
```

Comment:
Nit: Does this make sense ? (To avoid concatenation when we just pass 1 dataset without sampling, which is our basic current implementation) ?
 
`final_datasets[k] = v[0] if len(v) == 1 else datasets.concatenate_datasets(v)`